internal/providers/registry.go

```diff
@@ -1,52 +1,318 @@
 package providers
 
 import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
 	"sync"
+	"time"
 
+	"github.com/filecoin-project/storetheindex/config"
+	"github.com/filecoin-project/storetheindex/internal/providers/discovery"
+	"github.com/filecoin-project/storetheindex/internal/providers/policy"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
 	"github.com/libp2p/go-libp2p-core/peer"
-	ma "github.com/multiformats/go-multiaddr"
 )
 
+var log = logging.Logger("providers")
+
 type Registry struct {
-	rwmutex   sync.RWMutex
+	actions   chan func()
 	providers map[peer.ID]*ProviderInfo
+	policy    *policy.Policy
+
+	discovery    discovery.Discovery
+	discoTimes   map[string]time.Time
+	closeOnce    sync.Once
+	closed       chan struct{}
+	pendingDisco int
+
+	dstore datastore.Datastore
 }
 
+// ProviderInfo is an immutable data sturcture that holds information about a
+// provider.  A ProviderInfo instance is never modified, but rather a new one
+// is created to update its contents.  This means existing references remain
+// valid.
 type ProviderInfo struct {
-	Addresses []ma.Multiaddr
+	AddrInfo      peer.AddrInfo
+	FcAddress     string
+	LastIndex     cid.Cid
+	LastIndexTime time.Time
 }
 
-func NewRegistry() *Registry {
-	return &Registry{
+// NewRegistry creates a new provider registry, giving it a Discovery interface
+// and provider policy configuration.
+func NewRegistry(providerCfg config.Providers, disco discovery.Discovery, dstore datastore.Datastore) (*Registry, error) {
+	// Create policy from config
+	providerPolicy, err := policy.New(providerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &Registry{
```

Comment:
consider adding a worker pool for actions - it would be good to have a pool to know how many will be going on at once

Reply:
For limiting concurrency to some max number (workers in pool), or simply for tracking how many actions are running concurrently?

Running actions here is a substitute for mutex-locking the sections of code run as an action and allows the caller to decide whether or not to wait for the code to finish running.  So the maximum running actions are, the number of concurrent requests being processed + any discoveries in progress.  There will never be multiple discoveries for the same provider within the same `RediscoverWait` time.